### DISCUSSION TITLE
Thread safety issue

### DISCUSSION THREAD

**Discussion by claasz** [httprb/http#774] [General]
I'm a http.rb user for several years. In fact, it's the standard http lib in the company I work for.

To my surprise, I've now noticed that [it's not thread safe](https://github.com/httprb/http/wiki/Thread-Safety). While bugs can happen, I'm wondering why such a major issue doesn't seem to have any priority at all? #371 was opened in 2016, #558 was opened in 2019; the warning on the [wiki page](https://github.com/httprb/http/wiki/Thread-Safety) is from 2022.

For me, this is a show stopper and I have to remove http.rb from all my Rails apps which run in a Puma (i.e. multi threaded) setup. 

Does anybody have an explanation why this issue has no momentum?

**Comment by tarcieri** [comment]
The answer, like pretty much everything in open source, is that no one has done the work yet. This project is run by volunteers who contribute in their spare time. If you’d like a fix prioritized, the best way is to open a PR.

`HTTP::Client` was never intended to be thread safe. We also don’t provide an out-of-the-box thread safe connection pool, although that would be a nice contribution.

`HTTP::Options` was intended to be thread safe but that regressed in #7. #558 is the best tracking issue and includes proposed solutions no one has actually yet imemented.
